crates/bevy_reflect/src/type_info.rs

```diff
@@ -287,6 +290,199 @@ impl TypeInfo {
     impl_cast_method!(as_value: Value => ValueInfo);
 }
 
+/// The base representation of a Rust type.
+///
+/// When possible, it is recommended to use [`&'static TypeInfo`] instead of this
+/// as it provides more information as well as being smaller
+/// (since a reference only takes the same number of bytes as a `usize`).
+///
+/// However, where a static reference to [`TypeInfo`] is not possible,
+/// such as with trait objects and other types that can't implement [`Typed`],
+/// this type can be used instead.
+///
+/// It only requires that the type implements [`TypePath`].
+///
+/// And unlike [`TypeInfo`], this type implements [`Copy`], [`Eq`], and [`Hash`],
+/// making it useful as a key type.
+///
+/// It's especially helpful when compared to [`TypeId`] as it can provide the
+/// actual [type path] when debugging, while still having the same performance
+/// as hashing/comparing [`TypeId`] directly—at the cost of a little more memory.
+///
+/// # Examples
+///
+/// ```
+/// use bevy_reflect::{Type, TypePath};
+///
+/// fn assert_char<T: ?Sized + TypePath>(t: &T) -> Result<(), String> {
+///     let ty = Type::of::<T>();
+///     if Type::of::<char>() == ty {
+///         Ok(())
+///     } else {
+///         Err(format!("expected `char`, got `{}`", ty.path()))
+///     }
+/// }
+///
+/// assert_eq!(
+///     assert_char(&'a'),
+///     Ok(())
+/// );
+/// assert_eq!(
+///     assert_char(&String::from("Hello, world!")),
+///     Err(String::from("expected `char`, got `alloc::string::String`"))
+/// );
+/// ```
+///
+/// [`&'static TypeInfo`]: TypeInfo
+#[derive(Copy, Clone)]
+pub struct Type {
+    type_path_table: TypePathTable,
+    type_id: TypeId,
+}
+
+impl Type {
+    /// Create a new [`Type`] from a type that implements [`TypePath`].
+    pub fn of<T: TypePath + ?Sized>() -> Self {
+        Self {
+            type_path_table: TypePathTable::of::<T>(),
+            type_id: TypeId::of::<T>(),
+        }
+    }
+
+    /// Returns the [`TypeId`] of the type.
+    pub fn type_id(&self) -> TypeId {
```

Comment:
Ooh I like that. We already shorten `type_path` to just `path` so I think we can do the same here.

My one concern is that we already use `type_id` elsewhere, such as in the `TypeInfo` methods. Should those stay the same?

I think they can since we do it for `type_path` already. But I’m not too sure. Looks like we kinda have a mix 🤔